### PR TITLE
update UpdateSell action

### PR DESCRIPTION
**Pull request by tyrosine1153**
### Description

- Update UpdateSell action to update sell item in one bundle
  -> It didn't create a new action, but improved the existing action.

### How to test

Modifying the `ExpirationInterval` value in [Order.cs](https://github.com/planetarium/lib9c/blob/development/Lib9c/Model/Order/Order.cs) helps to test UpdateSell action .

### Related Links

[샵 개선: 일괄 재등록 lib9c](https://app.asana.com/0/958521740385861/1202166940058283/f)

### CODE REVIEW

File: Lib9c/Action/UpdateSell.cs [hunk 1]
```diff
@@ -26,38 +25,25 @@ namespace Nekoyume.Action
     /// Updated at https://github.com/planetarium/lib9c/pull/957
```
Comment (sonohoshi) [inline Lib9c/Action/UpdateSell.cs:25]:
```suggestion
    /// Hard forked at https://github.com/planetarium/lib9c/pull/1022
    /// Updated at https://github.com/planetarium/lib9c/pull/1022
```

**Review by hskim881028 — DISMISSED** [review]

File: Lib9c/Action/UpdateSell.cs (hunk 1 above)
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs:25]: Can I know what this comment means?

File: Lib9c/Action/UpdateSell.cs [hunk 2]
```diff
@@ -93,119 +65,133 @@ public override IAccountStateDelta Execute(IActionContext context)
...
-            if (!avatarState.worldInformation.IsStageCleared(
-                GameConfig.RequireClearedStageLevel.ActionsInShop))
+            if (!avatarState.worldInformation.IsStageCleared(GameConfig.RequireClearedStageLevel.ActionsInShop))
             {
                 avatarState.worldInformation.TryGetLastClearedStageId(out var current);
-                throw new NotEnoughClearedStageLevelException(
-                    addressesHex,
-                    GameConfig.RequireClearedStageLevel.ActionsInShop,
-                    current);
+                throw new NotEnoughClearedStageLevelException(addressesHex,
```
[the remark this answers is not in the document]
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs:84]: There's no change in the content, should I change it again?

File: Lib9c/Action/UpdateSell.cs [hunk 3]
```diff
@@ -118,94 +88,120 @@ public override IAccountStateDelta Execute(IActionContext context)
...
+                var updateSellShopState = states.TryGetState(updateSellShopAddress, out Dictionary serializedState)
+                    ? new ShardedShopStateV2(serializedState)
+                    : new ShardedShopStateV2(updateSellShopAddress);
+
+                Log.Verbose("{AddressesHex} UpdateSell Get ShardedShopState: {Elapsed}", addressesHex, sw.Elapsed);
+                sw.Restart();
+                var newOrder = OrderFactory.Create(context.Signer, sellerAvatarAddress,
+                    updateSellInfo.updateSellOrderId, updateSellInfo.price,
+                    updateSellInfo.tradableId, context.BlockIndex, updateSellInfo.itemSubType, updateSellInfo.count);
+                newOrder.Validate(avatarState, updateSellInfo.count);
```
Comment (ipdae) [inline Lib9c/Action/UpdateSell.cs:179]: Validate throw Exception. we must handle this.

**Review by ipdae — CHANGES_REQUESTED** [review]
It would be better to check all validation first. i think it to be a risk of a case in which the state changes in the middle if processing is performed in the existing order.

File: Lib9c/Action/UpdateSell2.cs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+using BxList = Bencodex.Types.List;
+
+namespace Nekoyume.Action
+{
+    /// <summary>
+    /// Hard forked at https://github.com/planetarium/lib9c/pull/602
+    /// Updated at https://github.com/planetarium/lib9c/pull/620
+    /// Updated at https://github.com/planetarium/lib9c/pull/861
+    /// Updated at https://github.com/planetarium/lib9c/pull/957
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell2.cs:27]: I fixed it, thank you!

File: Lib9c/Action/UpdateSell.cs [hunk 5]
```diff
@@ -93,119 +65,133 @@ public override IAccountStateDelta Execute(IActionContext context)
...
 
-            var expirationMail = avatarState.mailBox.OfType<OrderExpirationMail>()
-                            .FirstOrDefault(m => m.OrderId.Equals(orderId));
-            if (!(expirationMail is null))
+            foreach (var updateSellInfo in updateSellInfos)
             {
-                avatarState.mailBox.Remove(expirationMail);
+                if (!states.TryGetState(digestListAddress, out Dictionary rawList))
+                {
+                    errors.Add((updateSellInfo.orderId, ShopErrorType.ERROR_CODE_FAILED_LOADING_STATE));
```
[the remark this answers is not in the document]
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs]: To apply this comment, I added tuple element, state address in error list.
Is this change in line with what you commented?

File: Lib9c/Action/UpdateSellInfo.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            orderId = serialized[OrderIdKey].ToGuid();
+            updateSellOrderId = serialized[updateSellOrderIdKey].ToGuid();
+            tradableId = serialized[ItemIdKey].ToGuid();
+            itemSubType = serialized[ItemSubTypeKey].ToEnum<ItemSubType>();
+            price = serialized[PriceKey].ToFungibleAssetValue();
+            count = serialized[ItemCountKey].ToInteger();
+        }
+
+        public IValue Serialize()
```
[the remark this answers is not in the document]
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSellInfo.cs:47]: I fixed it. How about this change?

File: Lib9c/Action/UpdateSell.cs (hunk 3 above)
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs]: I fixed it. But I'm not sure this change was valid. Could you check it again?

File: Lib9c/Action/UpdateSell.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs]: Sorry, All parts related to the review have been deleted. Could you check it again?

File: Lib9c/Action/UpdateSell.cs (hunk 3 above)
Comment (tyrosine1153) [reply Lib9c/Action/UpdateSell.cs]: Sorry, All parts related to the review have been deleted. Could you check it again?

File: Lib9c/Action/UpdateSell.cs [hunk 7]
```diff
@@ -20,44 +19,27 @@
...
         protected override void LoadPlainValueInternal(IImmutableDictionary<string, IValue> plainValue)
         {
-            orderId = plainValue[OrderIdKey].ToGuid();
-            updateSellOrderId = plainValue[updateSellOrderIdKey].ToGuid();
-            tradableId = plainValue[ItemIdKey].ToGuid();
             sellerAvatarAddress = plainValue[SellerAvatarAddressKey].ToAddress();
-            itemSubType = plainValue[ItemSubTypeKey].ToEnum<ItemSubType>();
-            price = plainValue[PriceKey].ToFungibleAssetValue();
-            count = plainValue[ItemCountKey].ToInteger();
+            updateSellInfos = plainValue[UpdateSellInfoKey].ToList(info => new UpdateSellInfo((List) info));
```
Comment (sonohoshi) [inline Lib9c/Action/UpdateSell.cs:42]:
```suggestion
            updateSellInfos = plainValue[UpdateSellInfoKey].ToEnumerable(info => new UpdateSellInfo((List)info));
```
not `List` !

**Review by sonohoshi — APPROVED** [review]

**Pull request merged**